File: python/src/falcon_core/physics/device_structures/connections.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.connections import Connections as _CConnections
from falcon_core.physics.device_structures.connection import Connection
from falcon_core.generic.list import List

class Connections:
    """Python wrapper for Connections."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new_empty(cls, ) -> Connections:
        return cls(_CConnections.new_empty())

    @classmethod
    def new(cls, items: List) -> Connections:
        obj = cls(_CConnections.new(items._c if items is not None else None))
        obj._ref_items = items  # Keep reference alive
        return obj

    @classmethod
    def from_json(cls, json: str) -> Connections:
        return cls(_CConnections.from_json(json))

    def is_gates(self, ) -> None:
        ret = self._c.is_gates()
        return ret

    def is_ohmics(self, ) -> None:
        ret = self._c.is_ohmics()
        return ret

    def is_dot_gates(self, ) -> None:
        ret = self._c.is_dot_gates()
        return ret

    def is_plunger_gates(self, ) -> None:
        ret = self._c.is_plunger_gates()
        return ret

    def is_barrier_gates(self, ) -> None:
        ret = self._c.is_barrier_gates()
        return ret

    def is_reservoir_gates(self, ) -> None:
        ret = self._c.is_reservoir_gates()
        return ret

    def is_screening_gates(self, ) -> None:
        ret = self._c.is_screening_gates()
        return ret

    def intersection(self, other: Connections) -> Connections:
        ret = self._c.intersection(other._c if other is not None else None)
        return Connections._from_capi(ret)

    def push_back(self, value: Connection) -> None:
        ret = self._c.push_back(value._c if value is not None else None)
        return ret

    def size(self, ) -> None:
        ret = self._c.size()
        return ret

    def empty(self, ) -> None:
        ret = self._c.empty()
        return ret

    def erase_at(self, idx: Any) -> None:
        ret = self._c.erase_at(idx)
        return ret

    def clear(self, ) -> None:
        ret = self._c.clear()
        return ret

    def at(self, idx: Any) -> Connection:
        ret = self._c.at(idx)
        if ret is None: return None
        return Connection._from_capi(ret)

    def items(self, ) -> List:
        ret = self._c.items()
        if ret is None: return None
        return List(ret)

    def contains(self, value: Connection) -> None:
        ret = self._c.contains(value._c if value is not None else None)
        return ret

    def index(self, value: Connection) -> None:
        ret = self._c.index(value._c if value is not None else None)
        return ret

    def equal(self, other: Connections) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: Connections) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def __len__(self):
        return self.size()

    def __getitem__(self, idx):
        ret = self.at(idx)
        if ret is None:
            raise IndexError("Index out of bounds")
        return ret

    def append(self, value):
        return self.push_back(value)

    @classmethod
    def from_list(cls, items):
        return cls(_CConnections.from_list(items))

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, Connections):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, Connections):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/physics/device_structures/connections.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Generic
open Falcon_core.Physics.Device_structures

class c_connections (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.connections_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module Connections = struct
  type t = c_connections

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.connections_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connections ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.connections_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_connections ptr

  let empty () : t =
    let ptr = Capi_bindings.connections_create_empty () in
    ErrorHandling.raise_if_error ();
    new c_connections ptr

  let make (items : ListConnection.t) : t =
    ErrorHandling.read items (fun () ->
      let ptr = Capi_bindings.connections_create items#raw in
      ErrorHandling.raise_if_error ();
      new c_connections ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.connections_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.connections_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let isGates (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_is_gates handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isOhmics (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_is_ohmics handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isDotGates (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_is_dot_gates handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isPlungerGates (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_is_plunger_gates handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isBarrierGates (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_is_barrier_gates handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isReservoirGates (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_is_reservoir_gates handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let isScreeningGates (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_is_screening_gates handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let intersection (handle : t) (other : t) : t =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.connections_intersection handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      new c_connections result
    )

  let pushBack (handle : t) (value : Connection.t) : unit =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.connections_push_back handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let size (handle : t) : int =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_size handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let empty (handle : t) : bool =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_empty handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let eraseAt (handle : t) (idx : int) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_erase_at handle#raw (Unsigned.Size_t.of_int idx) in
      ErrorHandling.raise_if_error ();
      result
    )

  let clear (handle : t) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_clear handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let at (handle : t) (idx : int) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_at handle#raw (Unsigned.Size_t.of_int idx) in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let items (handle : t) : ListConnection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.connections_items handle#raw in
      ErrorHandling.raise_if_error ();
      new c_listconnection result
    )

  let contains (handle : t) (value : Connection.t) : bool =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.connections_contains handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let index (handle : t) (value : Connection.t) : int =
    ErrorHandling.multi_read [handle; value] (fun () ->
      let result = Capi_bindings.connections_index handle#raw value#raw in
      ErrorHandling.raise_if_error ();
      result
    )

end